knative-operator/pkg/controller/knativekafka/knativekafka_controller.go

```diff
@@ -280,29 +320,41 @@ func (r *ReconcileKnativeKafka) delete(instance *operatorv1alpha1.KnativeKafka)
 }
 
 func (r *ReconcileKnativeKafka) deleteKnativeKafka(instance *operatorv1alpha1.KnativeKafka) error {
-	manifest, err := r.buildManifest(instance)
+	manifest, err := r.buildManifest(instance, ManifestBuildAll)
 	if err != nil {
 		return fmt.Errorf("failed to build manifest: %w", err)
 	}
 
-	if err := r.transform(manifest, instance); err != nil {
-		return fmt.Errorf("failed to transform manifest: %w", err)
+	stages := []func(*mf.Manifest, *operatorv1alpha1.KnativeKafka) error{
+		r.transform,
+		r.deleteResources,
 	}
 
-	if err := manifest.Delete(); err != nil {
-		return fmt.Errorf("failed to delete KnativeKafka manifest: %w", err)
+	// Execute each stage in sequence until one returns an error
+	for _, stage := range stages {
+		if err := stage(manifest, instance); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (r *ReconcileKnativeKafka) buildManifest(instance *operatorv1alpha1.KnativeKafka) (*mf.Manifest, error) {
+type manifestBuild int
+
+const (
+	ManifestBuildEnabledOnly manifestBuild = iota
+	ManifestBuildDisabledOnly
+	ManifestBuildAll
+)
+
+func (r *ReconcileKnativeKafka) buildManifest(instance *operatorv1alpha1.KnativeKafka, build manifestBuild) (*mf.Manifest, error) {
 	var resources []unstructured.Unstructured
 
-	if instance.Spec.Channel.Enabled {
+	if build == ManifestBuildAll || (build == ManifestBuildEnabledOnly && instance.Spec.Channel.Enabled) || (build == ManifestBuildDisabledOnly && !instance.Spec.Channel.Enabled) {
```

Comment:
At this point, I think we should generalize this like so:

```golang
comps := map[mf.Manifest]bool{
  r.rawKafkaChannelManifest: instance.Spec.Channel.Enabled,
  r.rawKafkaSourceManifest: instance.Spec.Source.Enabled,
}

for manifest, enabled := range comps {
  	if build == ManifestBuildAll || (build == ManifestBuildEnabledOnly && enabled) || (build == ManifestBuildDisabledOnly && !enabled) {
		resources = append(resources, manifestResources()...)
	}
}
```

Reply:
yeah, why not. not a huge gain for 2 components, but sounds good.

Reply:
hmm, can't use `mf.Manifest` as map key. It doesn't have `==` and `!=`.

can convert the map into a slice of tuples...

any alternative thoughts?

Reply:
```golang
appendIfNecessary := func(enabled bool, manifest mf.Manifest) {
	if build == manifestBuildAll || (build == manifestBuildEnabledOnly && enabled) || (build == manifestBuildDisabledOnly && !enabled) {
		resources = append(resources, manifest.Resources()...)
	}
}

appendIfNecessary(instance.Spec.Channel.Enabled, r.rawKafkaChannelManifest)
appendIfNecessary(instance.Spec.Source.Enabled, r.rawKafkaSourceManifest)
```

Up to you if you like that better though. I mostly wouldn't want to repeat the big `if` condition.

Reply:
IMO, 2 explicit `if`s are easier to read than this. if we have a 3rd component  in the future we can convert it.

thanks for the reviews. if we're good, now let's wait until the other PR is merged.